sprint-1/step-2

Korotaew opened the pull request.

Korotaew committed:
sprint-1/step-2

Korotaew committed:
deleted unnecessary style

andryxxa93 commented on src/components/BurgerConstructor/BurgerConstructor.jsx:
```diff
@@ -0,0 +1,132 @@
+            <div className={styles.items}>
+                <div className={styles.bun}>
+                    <ConstructorElement
+                        type="top"
```
**Можно лучше**
Вы уже получаете данные с сервера, тут можно использовать их, а не моковые данные

andryxxa93 commented on src/components/BurgerIngredients/BurgerIngredients.jsx:
```diff
@@ -0,0 +1,96 @@
+  const [ingredients, setIngredients] = useState([]);
+
+  useEffect(() => {
+    const fetchData = async () => {
```
**Необходимо исправить**
Ингредиенты нужно получать только в одном месте, иначе у вас происходит дубликат вызова.

andryxxa93 commented on src/components/BurgerIngredients/BurgerIngredients.jsx:
```diff
@@ -0,0 +1,96 @@
+  };
+
+  BurgerIngredients.propTypes = {
+    data: PropTypes.arrayOf(ingredientTypes)
```
**Надо исправить**

в описании propTypes этого и других компонентов нужно указывать `.isRequired` у пропсов, без этого не контролируется, что пропс передается

Проверьте во всем проекте

andryxxa93 commented on src/components/IngredientDetails/IngredientDetails.jsx:
```diff
@@ -0,0 +1,58 @@
+
+function IngredientDetails({onClose, ingredient}) {
+    return (
+        <Modal onClose={onClose} isHeaderShow={true}>
```
**Необходимо исправить**
`Modal` необходимо вынести из компонентов `IngredientDetails ` и `OrderDetails`, чтобы их можно было использовать без модалки.
Должно быть так 
```
<Modal>
<IngredientDetails />
</Modal>
```

andryxxa93 commented on src/components/Modal/Modal.jsx:
```diff
@@ -0,0 +1,59 @@
+            closeModal();
+        }
+    }
+    useEffect(() => {
```
**Надо исправить**
Обработчик закрытия попапа должен добавляться при открытии попапа и удаляться при закрытии. Сейчас обработчик не будет удаляться при закрытии попапа т.к. из хука useEffect нужно возвращать колбэк отписки. Должно быть примерно так:

```
   useEffect(() => {
      const handleESCclose = () => {............} //обработчик события
      document.addEventListener("keydown", handleESCclose); //навешиваем обработчик
	  
	  //возвращаем колбэк отписки, который  будет вызван при размонтировании компонента
      return () => document.removeEventListener("keydown", handleESCclose)                                                                                               
   }, [onClose])
```

andryxxa93 commented on src/utils/api.js:
```diff
@@ -0,0 +1,8 @@
+export const getIngredients = () => {
+    return fetch("https://norma.nomoreparties.space/api/ingredients")
+      .then((res) => res.json())
```
**Надо исправить**
Добавьте проверку ответа сервера - res.ok:

Например:

```
.then(res => res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
```

иначе сейчас, если сервер ответи не кодом 2xx, а какой либо ошибкой (4xx, 5xx) мы не попадем в catch т.к. в этом случае fetch не выбрасывает ошибку, мы попытаемся отрисовать данные, хотя они не пришли, в итоге на попытке отрисовки несуществующих данных упадет все приложение

[https://learn.javascript.ru/fetch](https://learn.javascript.ru/fetch)

[https://developer.mozilla.org/en-US/docs/Web/API/Response/ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok)

Т.к. эта проверка нужна у всех запросов, лучше вынести её в отдельную функцию

```
const checkReponse = (res) => {
  return res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
};
```

`fetch(someUrl).then(checkReponse)`

andryxxa93 reviewed (changes requested):
Здравствуйте! Проделана большая работа 👍 

Что понравилось:
- Хорошая работа с `propTypes`
- Реализовали закрытие модалки по ESC 
- Проект соответствует макету

Однако есть пару моментов, которые необходимо исправить, чтобы работа была принята.

Необходимо исправить:
- На небольших экранах конструктор залезает на список 
 <img width="891" alt="Снимок экрана 2023-06-03 в 09 22 45" src="https://github.com/wety03/stellar-burger-sprint-1-step2/assets/60173362/8cf91ac5-1564-4b9a-8b31-4cd6bf6c95ba">

- Добавьте `isRequired` у пропсов
- обработчик закрытия попапа по ESC не будет удаляться при закрытии
- Добавьте проверку ответа сервера - res.ok

Более подробно описание проблемы смотрите комментарии в pull request.

В pull request вы так же найдете комментарии с рекомендациями как **можно улучшить** код. Их делать не обязательно, однако в будущем они могут помочь избежать ошибок и сделать код лучше.

**Напомню, чтобы работа была принята нужно исправить все замечания помеченные как **необходимо исправить**.

Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка F12) перед отправкой на ревью. 

Удачных правок!

Korotaew committed:
sprint-1/step-2 update

исправил большинство ошибок. Немного не понял о чем имелось ввиду в правках modal. Остальное все исправлено включая залезания конструктора на ингредиенты

Korotaew committed:
добавил пустой массив во второй аргумент useEffect

Korotaew committed:
Delete Modal.jsx

Korotaew committed:
Update Modal.jsx

передан useEffect пустой массив как и говорилось в правках

Korotaew committed:
Merge pull request #2 from wety03/sprint-1/step-2-update

sprint-1/step-2 update
